File: app/src/main/java/com/cyl/musiclake/ui/music/player/PlayControlFragment.java
package com.cyl.musiclake.ui.music.player;

import android.animation.ObjectAnimator;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.graphics.Palette;
import android.view.View;
import android.view.animation.LinearInterpolator;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import com.cyl.musiclake.R;
import com.cyl.musiclake.base.BaseFragment;
import com.cyl.musiclake.common.NavigationHelper;
import com.cyl.musiclake.common.TransitionAnimationUtils;
import com.cyl.musiclake.bean.Music;
import com.cyl.musiclake.event.MetaChangedEvent;
import com.cyl.musiclake.event.PlayModeEvent;
import com.cyl.musiclake.event.StatusChangedEvent;
import com.cyl.musiclake.player.FloatLyricViewManager;
import com.cyl.musiclake.player.PlayManager;
import com.cyl.musiclake.ui.UIUtils;
import com.cyl.musiclake.ui.music.playqueue.PlayQueueDialog;
import com.cyl.musiclake.utils.ColorUtil;
import com.cyl.musiclake.utils.CoverLoader;
import com.cyl.musiclake.utils.FormatUtil;
import com.cyl.musiclake.view.LyricView;
import com.cyl.musiclake.view.PlayPauseView;

import net.steamcrafted.materialiconlib.MaterialIconView;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.jetbrains.annotations.Nullable;

import butterknife.BindView;
import butterknife.OnClick;
import de.hdodenhof.circleimageview.CircleImageView;

public class PlayControlFragment extends BaseFragment<PlayPresenter> implements SeekBar.OnSeekBarChangeListener, PlayContract.View {

    private static final String TAG = "PlayControlFragment";
    public View topContainer;
    //整个容器
    @BindView(R.id.container)
    LinearLayout mContainer;
    @BindView(R.id.play_next)
    ImageButton mBtnNext;
    @BindView(R.id.song_progress_normal)
    ProgressBar mProgressBar;
    @BindView(R.id.play_pause)
    PlayPauseView mPlayPause;
    @BindView(R.id.title)
    TextView mTvTitle;
    @BindView(R.id.artist)
    TextView mTvArtist;
    @BindView(R.id.album)
    CircleImageView mIvAlbum;

    @BindView(R.id.skip_queue)
    MaterialIconView skip_queue;
    @BindView(R.id.previous)
    MaterialIconView skip_prev;
    @BindView(R.id.skip_next)
    MaterialIconView skip_next;
    @BindView(R.id.iv_love)
    public ImageView mIvLove;
    @BindView(R.id.iv_play_page_bg)
    ImageView ivPlayingBg;
    @BindView(R.id.playOrPause)
    PlayPauseView mPlayOrPause;
    @BindView(R.id.pb_loading)
    ProgressBar mLoadingPrepared;

    //textView
    @BindView(R.id.song_title)
    TextView mTvName;
    @BindView(R.id.song_artist)
    TextView mTvArtistAlbum;

    @BindView(R.id.song_elapsed_time)
    TextView tv_time;
    @BindView(R.id.song_duration)
    TextView tv_duration;

    @BindView(R.id.song_progress)
    SeekBar mSeekBar;
    @BindView(R.id.lyricView)
    LyricView mLrcView;

    private Palette mPalette;
    private Palette.Swatch mSwatch;
    private LinearInterpolator mLinearInterpolator = new LinearInterpolator();
    private ObjectAnimator coverAnimator;
    private long currentPlayTime = 0;

    @OnClick(R.id.skip_next)
    void next() {
        if (UIUtils.INSTANCE.isFastClick()) return;
        PlayManager.next();
    }

    @OnClick(R.id.play_next)
    void nextBottom() {
        if (UIUtils.INSTANCE.isFastClick()) return;
        PlayManager.next();
    }

    @OnClick(R.id.play_pause)
    void playOrPause() {
        PlayManager.playPause();
    }

    @OnClick(R.id.playOrPause)
    void playOrPauseF() {
        PlayManager.playPause();
    }

    @OnClick(R.id.previous)
    void prev() {
        if (UIUtils.INSTANCE.isFastClick()) return;
        PlayManager.prev();
    }

    @OnClick(R.id.iv_love)
    void love() {
        Music music = PlayManager.getPlayingMusic();
        if (music == null)
            return;
        UIUtils.INSTANCE.collectMusic(mIvLove, music);
    }


    @OnClick(R.id.skip_queue)
    void openPlayQueue() {
        PlayQueueDialog.Companion.newInstance().showIt((AppCompatActivity) mFragmentComponent.getActivity());
    }

    public static PlayControlFragment newInstance() {
        Bundle args = new Bundle();
        PlayControlFragment fragment = new PlayControlFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public int getLayoutId() {
        return R.layout.frag_player;
    }

    @Override
    public void initViews() {
        //初始化控件
        topContainer = rootView.findViewById(R.id.top_container);
        showLyric(FloatLyricViewManager.lyricInfo, true);
        updatePlayStatus(PlayManager.isPlaying());
    }

    @Override
    protected void initInjector() {
        mFragmentComponent.inject(this);
    }

    @Override
    protected void listener() {
        mSeekBar.setOnSeekBarChangeListener(this);
        topContainer.setOnClickListener(v -> {
            NavigationHelper.INSTANCE.navigateToPlaying(mFragmentComponent.getActivity(), mIvAlbum);
        });
    }

    @Override
    protected void loadData() {
        Music music = PlayManager.getPlayingMusic();
        if (mPresenter != null) {
            mPresenter.updateNowPlaying(music, true);
        }
        initAlbumPic(mIvAlbum);
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        if (PlayManager.isPlaying() || PlayManager.isPause()) {
            int progress = seekBar.getProgress();
            PlayManager.seekTo(progress);
            tv_time.setText(FormatUtil.INSTANCE.formatTime(progress));
            mLrcView.setCurrentTimeMillis(progress);
        } else {
            seekBar.setProgress(0);
        }
    }


    /**
     * 旋转动画
     */
    public void initAlbumPic(View view) {
        coverAnimator = ObjectAnimator.ofFloat(view, "rotation", 0, 359);
        coverAnimator.setDuration(20 * 1000);
        coverAnimator.setRepeatCount(-1);
        coverAnimator.setRepeatMode(ObjectAnimator.RESTART);
        coverAnimator.setInterpolator(mLinearInterpolator);
    }


//    @Override
//    public void setPalette(Palette palette) {
//        mPalette = palette;
//        mSwatch = ColorUtil.getMostPopulousSwatch(palette);
//
//        int paletteColor;
//        if (mSwatch != null) {
//            paletteColor = mSwatch.getRgb();
//            int artistColor = mSwatch.getTitleTextColor();
//            mTvName.setTextColor(ColorUtil.getOpaqueColor(artistColor));
//            mTvArtistAlbum.setTextColor(artistColor);
//        } else {
//            mSwatch = palette.getMutedSwatch() == null ? palette.getVibrantSwatch() : palette.getMutedSwatch();
//            if (mSwatch != null) {
//                paletteColor = mSwatch.getRgb();
//                int artistColor = mSwatch.getTitleTextColor();
//                mTvName.setTextColor(ColorUtil.getOpaqueColor(artistColor));
//                mTvArtistAlbum.setTextColor(artistColor);
//            } else {
//                paletteColor = Color.WHITE;
//                mTvName.setTextColor(ContextCompat.getColor(getContext(), android.R.color.primary_text_light));
//                mTvArtistAlbum.setTextColor(ContextCompat.getColor(getContext(), android.R.color.secondary_text_light));
//            }
//        }
//        //set icon color
//        int blackWhiteColor = ColorUtil.getBlackWhiteColor(paletteColor);
////        int statusBarColor = ColorUtil.getStatusBarColor(paletteColor);
////        StatusBarUtil.setColor(getActivity(), statusBarColor);
////        mLrcView.setHighLightTextColor(statusBarColor);
////        mLrcView.setDefaultColor(mSwatch.getBodyTextColor());
////        tv_time.setTextColor(blackWhiteColor);
////        mTvArtistAlbum.setTextColor(blackWhiteColor);
////        tv_duration.setTextColor(blackWhiteColor);
////        mLrcView.setHintColor(blackWhiteColor);
////        mBtnNext.setEnabled(true);
////        mBtnNext.setcolo(blackWhiteColor);
//        skip_prev.setColor(blackWhiteColor);
//        skip_next.setColor(blackWhiteColor);
//        skip_queue.setColor(blackWhiteColor);
//        mPlayOrPause.setBtnColor(blackWhiteColor);
//        mPlayOrPause.setEnabled(true);
//    }

    @Override
    public void showLyric(String lyricInfo, boolean isFilePath) {
        //初始化歌词配置
        mLrcView.setTouchable(false);
        mLrcView.setOnPlayerClickListener((progress, content) -> {
            PlayManager.seekTo((int) progress);
            if (!PlayManager.isPlaying()) {
                PlayManager.playPause();
            }
        });
        mLrcView.setLyricContent(lyricInfo);
    }


    @Override
    public void showLoading() {
        mLoadingPrepared.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        mLoadingPrepared.setVisibility(View.GONE);
    }


    @Override
    public void onPause() {
        super.onPause();
        if (coverAnimator != null) {
            coverAnimator.pause();
        }
    }

    @Override
    public void setPlayingBitmap(Bitmap albumArt) {

        if (coverAnimator != null) {
            if (PlayManager.isPlaying()) {
                coverAnimator.setCurrentPlayTime(currentPlayTime);
                coverAnimator.start();
            } else {
                coverAnimator.cancel();
                currentPlayTime = coverAnimator.getCurrentPlayTime();
            }
        }
    }

    @Override
    public void setPlayingBg(Drawable albumArt, Boolean isInit) {
        //加载背景图过度
        TransitionAnimationUtils.startChangeAnimation(ivPlayingBg, albumArt);
    }

    @Override
    public void updatePlayStatus(boolean isPlaying) {
        if (isPlaying) {
            mPlayPause.play();
            mPlayOrPause.play();
        } else {
            mPlayPause.pause();
            mPlayOrPause.pause();
        }
        if (coverAnimator != null) {
            if (isPlaying) {
                if (coverAnimator.isStarted()) {
                    coverAnimator.resume();
                } else {
                    coverAnimator.start();
                }
            } else {
                coverAnimator.pause();
            }
        }
    }

    @Override
    public void updatePlayMode() {

    }

    @Override
    public void updateProgress(long progress, long max) {
//        mSeekBar.setProgress((int) progress);
        mProgressBar.setProgress((int) progress);
//        tv_time.setText(FormatUtil.INSTANCE.formatTime(progress));
        mLrcView.setCurrentTimeMillis(progress);
//        mSeekBar.setMax((int) max);
        mProgressBar.setMax((int) max);
//        tv_duration.setText(FormatUtil.INSTANCE.formatTime(max));
    }

    @Override
    public void showNowPlaying(@Nullable Music music) {
        if (music != null) {
            rootView.setVisibility(View.VISIBLE);
            mTvName.setText(music.getTitle());
            mTvTitle.setText(music.getTitle());

            mTvArtist.setText(music.getArtist());
            mTvArtistAlbum.setText(music.getArtist());

            CoverLoader.loadImageView(getContext(),music.getCoverUri(),mIvAlbum);

            //更新收藏状态
            if (music.isLove()) {
                mIvLove.setImageResource(R.drawable.item_favorite_love);
            } else {
                mIvLove.setImageResource(R.drawable.item_favorite);
            }
        } else {
            rootView.setVisibility(View.GONE);
        }

    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onPlayModeChangedEvent(PlayModeEvent event) {
        updatePlayMode();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMetaChangedEvent(MetaChangedEvent event) {
        if (mPresenter != null) {
            mPresenter.updateNowPlaying(event.getMusic(), false);
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onStatusChangedEvent(StatusChangedEvent event) {
        mPlayPause.setLoading(!event.isPrepared());
        mPlayOrPause.setLoading(!event.isPrepared());
        updatePlayStatus(event.isPlaying());
    }

    @Override
    public void onResume() {
        super.onResume();
        if (coverAnimator != null && coverAnimator.isPaused() && PlayManager.isPlaying()) {
            coverAnimator.resume();
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        if (coverAnimator != null && coverAnimator.isRunning()) {
            coverAnimator.pause();
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EventBus.getDefault().register(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        topContainer = null;
        coverAnimator = null;
        EventBus.getDefault().unregister(this);
    }

}

File: app/src/main/java/com/cyl/musiclake/ui/music/bottom/PlayControlFragment.kt
package com.cyl.musiclake.ui.music.bottom

import android.animation.ObjectAnimator
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.SeekBar
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.PagerSnapHelper
import androidx.recyclerview.widget.RecyclerView
import com.cyl.musiclake.R
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.common.NavigationHelper
import com.cyl.musiclake.event.MetaChangedEvent
import com.cyl.musiclake.event.PlayModeEvent
import com.cyl.musiclake.event.PlaylistEvent
import com.cyl.musiclake.event.StatusChangedEvent
import com.cyl.musiclake.ui.base.BaseFragment
import com.cyl.musiclake.ui.music.playpage.PlayContract
import com.cyl.musiclake.ui.music.playpage.PlayPresenter
import com.cyl.musiclake.ui.music.playqueue.PlayQueueDialog
import com.cyl.musiclake.utils.LogUtil
import com.music.lake.musiclib.MusicPlayerManager
import com.music.lake.musiclib.bean.BaseMusicInfo
import kotlinx.android.synthetic.main.play_control_menu.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.jetbrains.anko.support.v4.runOnUiThread
import java.util.*
import kotlin.math.max
import kotlin.math.min

class PlayControlFragment : BaseFragment<PlayPresenter>(), SeekBar.OnSeekBarChangeListener, PlayContract.View {

    private var coverAnimator: ObjectAnimator? = null
    private var mAdapter: BottomMusicAdapter? = null
    private val musicList = ArrayList<BaseMusicInfo>()

    override fun getLayoutId(): Int {
        return R.layout.play_control_menu
    }

    public override fun initViews() {
        //初始化控件
        updatePlayStatus(MusicPlayerManager.getInstance().isPlaying)
        musicList.clear()
        if (MusicPlayerManager.getInstance().playList != null) {
            musicList.addAll(MusicPlayerManager.getInstance().playList)
        }
        initSongList()
    }

    override fun initInjector() {
        mFragmentComponent.inject(this)
    }

    override fun listener() {
        bottomPlayRcv.setOnClickListener { v -> NavigationHelper.navigateToPlaying(mFragmentComponent.activity) }
        playQueueIv.setOnClickListener {
            PlayQueueDialog.newInstance().showIt((activity as AppCompatActivity))
        }
        playPauseView.setOnClickListener {
            MusicPlayerManager.getInstance().pausePlay()
        }
    }

    override fun loadData() {
        val music = MusicPlayerManager.getInstance().getNowPlayingMusic()
        music?.let {
            mPresenter?.updateNowPlaying(it, true)
        }
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {

    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {

    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        if (MusicPlayerManager.getInstance().isPlaying()) {
            val progress = seekBar.progress
            MusicPlayerManager.getInstance().seekTo(progress.toLong())
        } else {
            seekBar.progress = 0
        }
    }


    override fun updatePlayStatus(isPlaying: Boolean) {
        if (isPlaying && !playPauseView.isPlaying) {
            playPauseView.play()
        } else if (!isPlaying && playPauseView.isPlaying) {
            playPauseView.pause()
        }
    }

    override fun updateLoading(isLoading: Boolean) {
        playPauseView.setLoading(isLoading)
    }

    override fun updateProgress(progress: Long, max: Long, bufferPercent: Int) {
        runOnUiThread {
            progressBar.progress = if (max <= 0) 0 else max(0, min((progress * 100 / max).toInt(), 100))
            progressBar.secondaryProgress = bufferPercent
            progressBar.max = 100
            playPauseView.setProgress(1.0f * progress / max)
        }
    }

    override fun showNowPlaying(baseMusicInfo: BaseMusicInfo?) {
        if (baseMusicInfo != null) {
            rootView.visibility = View.VISIBLE
        } else {
            rootView.visibility = View.GONE
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onPlayModeChangedEvent(event: PlayModeEvent) {
        updatePlayMode()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMetaChangedEvent(event: MetaChangedEvent) {
        mPresenter?.updateNowPlaying(event.baseMusicInfoInfo, false)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            progressBar.setProgress(MusicPlayerManager.getInstance().getPlayingPosition().toInt(), true)
        } else {
            progressBar.progress = MusicPlayerManager.getInstance().getPlayingPosition().toInt()
        }
        progressBar.max = MusicPlayerManager.getInstance().getDuration().toInt()
        bottomPlayRcv.scrollToPosition(MusicPlayerManager.getInstance().getNowPlayingIndex())
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onStatusChangedEvent(event: StatusChangedEvent) {
        updatePlayStatus(event.isPlaying)
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onPlayListChange(event: PlaylistEvent) {
        if (event.type == Constants.PLAYLIST_QUEUE_ID) {
            val playlist = MusicPlayerManager.getInstance().playList
            val index = MusicPlayerManager.getInstance().nowPlayingIndex
            LogUtil.d(TAG, "播放列表已改变" + playlist.size + " - " + index)
            musicList.clear()
            musicList.addAll(playlist)
            mAdapter?.notifyItemRangeChanged(0, playlist.size)
            bottomPlayRcv.scrollToPosition(index)
            if (musicList.size == 0) {
                emptyView.visibility = View.VISIBLE
            } else {
                emptyView.visibility = View.GONE
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        EventBus.getDefault().register(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        coverAnimator = null
        EventBus.getDefault().unregister(this)
    }

    /**
     * 初始化歌曲列表
     */
    private fun initSongList() {
        if (mAdapter == null) {
            bottomPlayRcv.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
            mAdapter = BottomMusicAdapter(musicList)
            val snap = PagerSnapHelper()
            snap.attachToRecyclerView(bottomPlayRcv)
            bottomPlayRcv.adapter = mAdapter
            bottomPlayRcv.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                    if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                        val manager = recyclerView?.layoutManager as LinearLayoutManager
                        val first = manager.findFirstVisibleItemPosition()
                        val last = manager.findLastVisibleItemPosition()
                        LogUtil.e("Scroll", "$first-$last")
                        if (first == last && first != MusicPlayerManager.getInstance().getNowPlayingIndex()) {
                            MusicPlayerManager.getInstance().playMusicById(first)
                        }
                    }
                }
            })
            mAdapter?.bindToRecyclerView(bottomPlayRcv)
        } else {
            mAdapter?.notifyDataSetChanged()
        }
        bottomPlayRcv.scrollToPosition(MusicPlayerManager.getInstance().getNowPlayingIndex())
        if (musicList.size == 0) {
            emptyView.visibility = View.VISIBLE
        } else {
            emptyView.visibility = View.GONE
        }
    }

    companion object {

        private val TAG = "PlayControlFragment"

        fun newInstance(): PlayControlFragment {
            val args = Bundle()
            val fragment = PlayControlFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
